storage/namespace_readers.go

```diff
@@ -0,0 +1,313 @@
+		return openReader.reader, nil
+	}
+
+	// Cache miss, need to return a reused reader or open a new reader
+	var (
+		reader fs.FileSetReader
+		err    error
+	)
+	if len(m.closedReaders) > 0 {
+		idx := len(m.closedReaders) - 1
+		reader = m.closedReaders[idx].reader
+		// Zero refs from element in slice and shrink slice
+		m.closedReaders[idx] = cachedReader{}
+		m.closedReaders = m.closedReaders[:idx]
+
+		m.metrics.cacheMissReusedReader.Inc(1)
+	} else {
+		reader, err = m.newReaderFn(m.bytesPool, m.fsOpts)
+		if err != nil {
+			return nil, err
+		}
+
+		m.metrics.cacheMissAllocReader.Inc(1)
+	}
+	// Can release the lock now as we prepare reader for caller
+	m.Unlock()
+
+	if err := reader.Open(m.namespace.ID(), shard, blockStart); err != nil {
+		return nil, err
+	}
+	// We can validate metadata immediately since its read when opened
+	if err := reader.ValidateMetadata(); err != nil {
+		return nil, err
+	}
+
+	// Fast fwd through if in the middle of a volume
+	for i := 0; i < position.dataIdx; i++ {
+		id, data, _, err := reader.Read()
+		if err != nil {
+			return nil, err
+		}
+		id.Finalize()
+		data.Finalize()
+	}
+	for i := 0; i < position.metadataIdx; i++ {
+		id, _, _, err := reader.ReadMetadata()
+		if err != nil {
+			return nil, err
+		}
+		id.Finalize()
+	}
+
+	return reader, nil
+}
+
+func (m *namespaceReaderManager) put(reader fs.FileSetReader) {
+	status := reader.Status()
+
+	m.Lock()
+	defer m.Unlock()
+
+	if !status.Open {
+		m.closedReaders = append(m.closedReaders, cachedReader{
+			reader: reader,
+		})
+		return
+	}
+
+	key := cachedOpenReaderKey{
+		shard:      status.Shard,
+		blockStart: xtime.ToUnixNano(status.BlockStart),
+		position: readerPosition{
+			dataIdx:     reader.EntriesRead(),
+			metadataIdx: reader.MetadataRead(),
+		},
+	}
+
+	if _, ok := m.openReaders[key]; ok {
+		// Unlikely, however if so just close the reader we were trying to put
+		// and put into the closed readers
+		if err := reader.Close(); err != nil {
+			m.logger.Errorf("error closing reader on put from reader cache: %v", err)
+			return
+		}
+		m.closedReaders = append(m.closedReaders, cachedReader{
+			reader: reader,
+		})
+		return
+	}
+
+	m.openReaders[key] = cachedReader{
+		reader: reader,
+	}
+}
+
+func (m *namespaceReaderManager) tick() {
+	m.tickWithThreshold(expireCachedReadersAfterNumTicks)
+}
+
+func (m *namespaceReaderManager) close() {
+	// Perform a tick but make the threshold zero so all readers must be expired
+	m.tickWithThreshold(0)
+}
+
+func (m *namespaceReaderManager) tickWithThreshold(threshold int) {
+	m.Lock()
+	defer m.Unlock()
+
+	// First increment ticks since used for closed readers
+	expiredClosedReaders := 0
+	for i := range m.closedReaders {
+		m.closedReaders[i].ticksSinceUsed++
+		if m.closedReaders[i].ticksSinceUsed >= threshold {
+			expiredClosedReaders++
+		}
+	}
+	// Expire any closed readers, alloc a new slice to avoid spikes
+	// of use creating slices that are never released
```

Comment:
Not sure what you're worried about here in terms of allocation, how would you create slices that are never released?

Reply:
Also I'm not sure why you need to expire the closed readers. The get/put logic should handle re-using them for you automatically right? Are you just trying to prevent the slice size from growing too large after a burst and never shrinking?

Reply:
The latter yes, preventing the slice from growing large after a burst and never releasing the memory back.